x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/X-IrateAggregator.java.st

```diff
@@ -0,0 +1,276 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.compute.aggregation;
+
+// begin generated imports
+import org.apache.lucene.util.Accountable;
+import org.apache.lucene.util.RamUsageEstimator;
+import org.elasticsearch.common.breaker.CircuitBreaker;
+import org.elasticsearch.common.util.BigArrays;
+import org.elasticsearch.common.util.ObjectArray;
+import org.elasticsearch.compute.ann.GroupingAggregator;
+import org.elasticsearch.compute.ann.IntermediateState;
+import org.elasticsearch.compute.data.Block;
+import org.elasticsearch.compute.data.BlockFactory;
+import org.elasticsearch.compute.data.DoubleBlock;
+import org.elasticsearch.compute.data.DoubleVector;
+import org.elasticsearch.compute.data.FloatBlock;
+import org.elasticsearch.compute.data.IntBlock;
+import org.elasticsearch.compute.data.IntVector;
+import org.elasticsearch.compute.data.LongBlock;
+import org.elasticsearch.compute.operator.DriverContext;
+import org.elasticsearch.core.Releasable;
+import org.elasticsearch.core.Releasables;
+// end generated imports
+
+/**
+ * A rate grouping aggregation definition for $type$.
+ * This class is generated. Edit `X-RateAggregator.java.st` instead.
+ */
+@GroupingAggregator(
+    value = { @IntermediateState(name = "timestamps", type = "LONG_BLOCK"), @IntermediateState(name = "values", type = "$TYPE$_BLOCK") }
+)
+public class Irate$Type$Aggregator {
+    public static $Type$IrateGroupingState initGrouping(DriverContext driverContext) {
+        return new $Type$IrateGroupingState(driverContext.bigArrays(), driverContext.breaker());
+    }
+
+    public static void combine($Type$IrateGroupingState current, int groupId, $type$ value, long timestamp) {
+        current.append(groupId, timestamp, value);
+    }
+
+    public static void combineIntermediate(
+        $Type$IrateGroupingState current,
+        int groupId,
+        LongBlock timestamps,
+        $Type$Block values,
+        int otherPosition
+    ) {
+        current.combine(groupId, timestamps, values, otherPosition);
+    }
+
+    public static Block evaluateFinal($Type$IrateGroupingState state, IntVector selected, GroupingAggregatorEvaluationContext evalContext) {
+        return state.evaluateFinal(selected, evalContext);
+    }
+
+    private static class $Type$IrateState {
+        static final long BASE_RAM_USAGE = RamUsageEstimator.sizeOfObject($Type$IrateState.class);
+        final long[] timestamps; // descending order
```

Comment:
Can we use 4 fields and a boolean/integer instead of two arrays? The state would be simpler.